compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1113,12 +1114,27 @@ public void visit(BLangTypeInit cIExpr) {
                 break;
             case TypeTags.STREAM:
             case TypeTags.CHANNEL:
-                if (cIExpr.initInvocation.argExprs.size() > 0) {
+                if (!cIExpr.initInvocation.argExprs.isEmpty()) {
                     dlog.error(cIExpr.pos, DiagnosticCode.TOO_MANY_ARGS_FUNC_CALL, cIExpr.initInvocation.name);
                     resultType = symTable.semanticError;
                     return;
                 }
                 break;
+            case TypeTags.UNION:
+                List<BType> matchingMembers = findMembersWithMatchingInitFunc(cIExpr, (BUnionType) actualType);
+                BType matchedType = getMatchedTypeOrEmitError(matchingMembers, cIExpr, actualType);
+
+                expType = matchedType;
+                if (matchedType.tsymbol.type.tag == TypeTags.OBJECT
```

Comment:
This can simply be `matchedType.tag`. Getting the type from the type symbol will again return the `matchedType`.